http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ResponseHandler.java

```diff
@@ -106,14 +109,52 @@ protected void channelRead0(ChannelHandlerContext channelContext, HttpObject msg
 
                 fullContent = ((FullHttpResponse) msg).content();
                 ByteBuffer bb = copyToByteBuffer(fullContent);
-                requestContext.handler().onStream(new FullResponseContentPublisher(channelContext, bb, ef));
-                finalizeResponse(requestContext, channelContext);
+                requestContext.handler().onStream(new DataCountingPublisher(channelContext,
+                                                                            new FullResponseContentPublisher(channelContext,
+                                                                                                             bb, ef)));
+
+                Optional<IOException> contentLengthFailure = validateResponseContentLength(channelContext);
+                if (contentLengthFailure.isPresent()) {
+                    exceptionCaught(channelContext, contentLengthFailure.get());
+                } else {
+                    finalizeResponse(requestContext, channelContext);
+                }
             } finally {
                 Optional.ofNullable(fullContent).ifPresent(ByteBuf::release);
             }
         }
     }
 
+    private Long responseContentLength(HttpResponse response) {
+        String length = response.headers().get("content-length");
+        if (length == null) {
+            return null;
+        }
+
+        return Long.parseLong(length);
+    }
+
+    private static Optional<IOException> validateResponseContentLength(ChannelHandlerContext ctx) {
+        Long contentLengthHeader = ctx.channel().attr(RESPONSE_CONTENT_LENGTH).get();
+        Long actualContentLength = ctx.channel().attr(RESPONSE_DATA_READ).get();
+
+        if (contentLengthHeader == null) {
+            return Optional.empty();
+        }
+
+        if (actualContentLength == null) {
+            actualContentLength = 0L;
+        }
+
+        if (actualContentLength.equals(contentLengthHeader)) {
+            return Optional.empty();
+        }
+
+        return Optional.of(new IOException("Response had content-length of " + contentLengthHeader + " bytes, but only received "
+                                           + actualContentLength + " before the connection was closed."));
```

Comment:
What if `Transfer-Encoding: chunked` is being used, but we don't receive the final (0 byte) chunk?

Reply:
Also what happens if an HTTP/2 channel is closed without an `endOfStream` flag? Does Netty throw an exception?

Reply:
That sounds like a different edge case we might need to add validation for. That's much rarer, since we rarely use chunked encoding. This particular code only goes into effect when the content-length is specified.

Reply:
Yeah, maybe we also need to test the end-of-stream flag as well.